grammar.js

```diff
@@ -251,15 +246,18 @@ module.exports = grammar({
 
             seq('(', $.expr, ')'),
             seq('copy', '(', $.expr, ')'),
-            prec_r(seq('hook', $.expr)),
-            seq($.expr, '?$', $.id),
+            prec_r(seq('hook', $.id, '(', optional(list1($.expr, ',')), ')')),
+            $.field_check,
             seq('schedule', $.expr, '{', $.event_hdr, '}'),
             seq('function', $.begin_lambda, $.func_body),
 
             // Lower precedence here to favor local-variable statements
             prec_r(-1, seq('local', $.id, '=', $.expr)),
         ),
 
+        field_access: $ => prec_l(seq($.expr, '$', $.id)),
```

Comment:
I'll do the rename.

Out of curiosity, are you aware of tooling reflecting directly on node names? I wonder whether I am making my life too hard by enumerating named nodes by hand in my tree queries.

Reply:
Well ... in the script formatter I'm frequently taking the type of a tree node from its `.type` field, which in the Python bindings is a string like `field_access_expr`. There are also some basic properties that you can query for a node, like `is_named()`. I'm not sure I see how else I'd do it ... does that help you? I don't think I know enough about your tree queries. They're probably more tree-sitter-esque than what I'm doing. :-)

Reply:
Btw I also haven't explored tree-sitter's fields feature much: https://tree-sitter.github.io/tree-sitter/creating-parsers#using-fields ... in case that'd help you.

Reply:
> Well ... in the script formatter I'm frequently taking the type of a tree node from its .type field, which in the Python bindings is a string like field_access_expr. 

In the bindings I am using I have access to the [node kind](https://docs.rs/tree-sitter/latest/tree_sitter/struct.Node.html#method.kind) but haven't explored being too dynamic there. Maybe I will try this.

> Btw I also haven't explored tree-sitter's fields feature much: https://tree-sitter.github.io/tree-sitter/creating-parsers#using-fields ... in case that'd help you.

Field names become necessary if a node can have multiple children of the same kind (of which some could be optional); in that case one needs field names to be able to tell whether a certain field is present. I haven't really come across cases in this grammar where that was needed and I was not able to work around it; this is probably due to this grammar preferring to nest deeply.

> I'm not sure I see how else I'd do it ... does that help you? I don't think I know enough about your tree queries.

The way I extract declarations currently looks like this, https://github.com/bbannier/zeek-language-server/blob/c136d86fbb217038698e940c5b04f3daa09236d0/src/query.rs#L114. This works reasonably well since I made declarations look uniform, but I could instead also just use a bigger query which can match nodes of different shapes with some pruning. While tree queries feel "right", I also ran into extremely bad performance with them when I tried to also extra zeekygen comments with that same query (where performance became 100,000x worse). I'll probably be able to work around that [again](https://github.com/bbannier/zeek-language-server/blob/c136d86fbb217038698e940c5b04f3daa09236d0/src/query.rs#L392-L393).

I'll move this PR to draft and try to get everything I need done and performant.

Reply:
> Field names become necessary if a node can have multiple children of the same kind (of which some could be optional)

Ah, yes! This can be awkward right now in the formatter: when I get to an optional field, I check its `.type` value to see what I'm dealing with, and consume accordingly. I also rely on indexing by number in a bunch of places. Sounds like fields would make this cleaner.

> I'll move this PR to draft and try to get everything I need done and performant.

Okay! :+1: